kernel/context.ml

```diff
@@ -39,16 +39,23 @@ let eq_annot eq eqr {binder_name=na1;binder_relevance=r1} {binder_name=na2;binde
 let hash_annot h {binder_name=n;binder_relevance=r} =
   Hashset.Combine.combinesmall (Sorts.relevance_hash r) (h n)
 
-let map_annot f {binder_name=na;binder_relevance} =
-  {binder_name=f na;binder_relevance}
+let map_annot f ({binder_name=na;binder_relevance} as a) =
+  let na' = f na in
+  if na == na' then a else {binder_name=na';binder_relevance}
 
-let map_annot_relevance fr {binder_name=na;binder_relevance=r} =
-  {binder_name=na;binder_relevance=fr r}
+let map_annot_het f {binder_name=na;binder_relevance} =
+  let na' = f na in
+  {binder_name=na';binder_relevance}
 
-let make_annot x r = {binder_name=x;binder_relevance=r}
+let map_annot_relevance fr ({binder_name=na;binder_relevance=r} as a) =
+  let r' = fr r in
+  if r == r' then a else {binder_name=na;binder_relevance=r'}
```

Comment:
I wonder if it would be sane to do
```suggestion
  if r == Obj.magic r' then Obj.magic a else {binder_name=na;binder_relevance=r'}
```
(then no need for a _het variant)

Since record `pbinder_annot` has no mutable fields, values with physically equal fields should be equivalent even ignoring typing